Utils/hdinsight-node-common/src/com/microsoft/azure/hdinsight/common/AbfsUri.java

```diff
@@ -0,0 +1,120 @@
+    public static final Pattern HTTP_URI_PATTERN = Pattern.compile(
+            StoragePathInfo.AdlsGen2RestfulPathPattern, Pattern.CASE_INSENSITIVE);
+
+    private final URI rawUri;
+    private final LaterInit<String> fileSystem = new LaterInit<>();
+    private final LaterInit<String> accountName = new LaterInit<>();
+    private final LaterInit<String> subPath = new LaterInit<>();
+
+    private AbfsUri(URI rawUri) {
+        this.rawUri = rawUri;
+    }
+
+    public URI getRawUri() {
+        return rawUri;
+    }
+
+    public String getFileSystem() {
+        return fileSystem.get();
+    }
+
+    public String getAccountName() {
+        return accountName.get();
+    }
+
+    public String getSubPath() {
+        return subPath.get();
+    }
+
+    public URI getUri() {
+        return URI.create(String.format("abfs://%s@%s.dfs.core.windows.net%s",
+                getFileSystem(), getAccountName(), getSubPath()));
+    }
+
+    // get root URI ending without "/"
+    public URI getRootUri() {
+        return URI.create(String.format("abfs://%s@%s.dfs.core.windows.net", getFileSystem(), getAccountName()));
+    }
+
+    public URI getUrl() {
+        return URI.create(String.format("https://%s.dfs.core.windows.net/%s%s",
+                getAccountName(), getFileSystem(), getSubPath()));
+    }
+
+    // get restful root url ending without "/"
+    public URI getRootUrl() {
+        return URI.create(String.format("https://%s.dfs.core.windows.net/%s", getAccountName(), getFileSystem()));
+    }
+
+    // get subPath starting without "/" except when subPath is empty
+    public URI getDirectoryParam() {
+        return getSubPath().length() == 0 || getSubPath().equals("/")
+                ? URI.create("/")
+                : URI.create(getSubPath().substring(1));
+    }
+
+    public static AbfsUri parse(final String rawUri) {
+        Matcher matcher;
+        if (StringUtils.startsWithIgnoreCase(rawUri, "abfs")) {
+            matcher = ABFS_URI_PATTERN.matcher(rawUri);
+        } else if (StringUtils.startsWithIgnoreCase(rawUri, "http")) {
+            matcher = HTTP_URI_PATTERN.matcher(rawUri);
+        } else {
+            throw new UnknownFormatConversionException("Unsupported ADLS Gen2 URI Scheme: " + rawUri);
+        }
+
+        if (matcher.matches()) {
+            AbfsUri abfsUri = new AbfsUri(URI.create(rawUri));
+            abfsUri.accountName.set(matcher.group("accountName"));
+            abfsUri.fileSystem.set(matcher.group("fileSystem"));
+            abfsUri.subPath.set(matcher.group("subPath"));
+            return abfsUri;
+        }
+
+        throw new UnknownFormatConversionException("Unmatched ADLS Gen2 URI: " + rawUri);
+    }
+
+    public static boolean isType(final String uri) {
+        return ABFS_URI_PATTERN.matcher(uri).matches() || HTTP_URI_PATTERN.matcher(uri).matches();
+    }
+
+    @Override
+    public String toString() {
```

Comment:
Looks we can introduce an interface for `xxxUri` later. 😄

Reply:
won't fix in this iteration.

---
In reply to: [361568121](https://github.com/microsoft/azure-tools-for-java/pull/3885#discussion_r361568121) [](ancestors = 361568121)